### PR TITLE
ARROW-4930: [C++] Improve find_package() support

### PR DESCRIPTION
**Pull request by kou**
Backward incompatibilities:

  * find_package(Arrow) doesn't find Arrow Python.
    Users need to use find_package(ArrowPython).

  * find_package(ArrowCuda) doesn't work on not Windows (case
    sensitive environments). Use find_package(ArrowCUDA) instead.

Summary:

  * Create arrow_find_package() in FindArrow.cmake and use it in
    Find*.cmake.

  * arrow_find_package() is a function that finds an Arrow related
    package. It supports ARROW_HOME/PARQUET_HOME environment variable,
    CMake package configuration and pkg-config.

  * arrow_find_package() can find libraries installed under
    lib/x86_64-linux-gnu/ (on Debian), lib64/ (on CentOS), lib/,
    bin/ (on Windows).

    * We can remove workaround for ARROW-1090 in python/setup.py:
      https://issues.apache.org/jira/projects/ARROW/issues/ARROW-1090

  * Create Find*.cmake per component. It means that FindArrow.cmake
    doesn't find Arrow Python. Use find_package(ArrowPython) instead
    of find_package(Arrow).

    * Install Find*.cmake to be reusable.

    * Changes for Linux packages are for this.

  * Rename arrowConfig.cmake, arrowTarget.cmake and so on to
    ArrowConfig.cmake, ArrowTarget.cmake and so on to use the same
    package for FindArrow.cmake. arrow-config.cmake is added for
    backward compatibility. So users can use find_package(arrow)
    instead of find_package(Arrow).

  * ArrowConfig.cmake provides more useful variables such as
    ARROW_SO_VERSION and ARROW_FULL_SO_VERSION.

  * Rename FindArrowCuda.cmake to FindArrowCUDA.cmake because "CUDA"
    is the official name.

  * Use find_package_handle_standard_args() in Find*.cmake. It means
    that we can check required version by find_package(Arrow 0.15.0).

    * Changes for Parquet are for detecting Parquet version.

  * Use consistent variable names in Find*.cmake with keeping backward
    compatibility:

    * ${PACKAGE}_LIBS -> ${PACKAGE}_LIB_DIR: Because we use
      ${ARROW_INCLUDE_DIR} and CMake uses _DIR, _DIRECTORY, _PATH and
      so on.

    * ${PACKAGE}_SHARED_IMP_LIB -> ${PACKAGE}_IMPORT_LIB: Because
      CMake doesn't use SHARED for import library such as CMAKE_IMPORT_LIBRARY_SUFFIX.

  * Find*.cmake provides CMake targets such as arrow_shared and
    arrow_static even if it uses ARROW_HOME or pkg-config. Users can
    use them to link to Arrow.

  * Remove needless ABI_VERSION handling in python/CMakeLists.txt
    because ABI_VERSION just exists just for backward compatibility.
    We should use only SO_VERSION.

  * Rename "executable" to "plasma_store_server" in
    plasma.pc. "executable" is deprecated.

  * Add arrow-python-flight.pc.

Resolved issues:

  * ARROW-5021 [C++] Review hardcoded "lib" paths in
    Find$PACKAGE.cmake related to endogenous libraries
    https://issues.apache.org/jira/browse/ARROW-5021

Removed workarounds:

  * ARROW-1090 [Python] build_ext usability
    https://issues.apache.org/jira/projects/ARROW/issues/ARROW-1090

  * ARROW-5025 [Python][Packaging] wheel for Windows are broken
    https://issues.apache.org/jira/browse/ARROW-5025

  * ARROW-5026 [Python][Packaging] conda package on non Windows is broken
    https://issues.apache.org/jira/browse/ARROW-5026

### CODE REVIEW

**Comment by kou** [comment]
@ursabot crossbow submit -g conda -g wheel -g linux

**Comment by kou** [comment]
@ursabot crossbow submit -g conda -g wheel -g linux

**Comment by kou** [comment]
@ursabot crossbow submit ubuntu-xenial

**Comment by kou** [comment]
@ursabot crossbow submit ubuntu-xenial -g wheel

**Comment by kou** [comment]
@ursabot crossbow submit wheel-osx-cp37m

**Comment by kou** [comment]
@ursabot crossbow submit -g conda -g wheel -g linux

**Review by pitrou — COMMENTED** [review]
Kudos for doing this! Just a couple comments.

File: cpp/cmake_modules/FindArrow.cmake [hunk 1]
```diff
@@ -15,147 +15,365 @@
...
+
+macro(arrow_find_package_pkg_config)
+  pkg_check_modules(${prefix}_PC ${pkg_config_name})
+  if(${prefix}_PC_FOUND)
+    set(${prefix}_USE_PKG_CONFIG TRUE PARENT_SCOPE)
+
+    set(include_dir "${${prefix}_PC_INCLUDEDIR}")
+    set(lib_dir "${${prefix}_PC_LIBDIR}")
+    set(shared_lib_paths "${${prefix}_PC_LINK_LIBRARIES}")
+    list(GET shared_lib_paths 0 first_shared_lib_path)
```
[the remark this answers is not in the document]
Comment (kou) [reply cpp/cmake_modules/FindArrow.cmake]: Done.

**Review by kou — COMMENTED** [review]
Thanks for reviewing this large change!

I could fix some bugs by your comments.

I also found some name problems by adding comments. For example, I renamed `SHARED_IMPLIB` to `IMPORT_LIB` because CMake has `CMAKE_IMPORT_LIBRARY_SUFFIX` variable. `SHARED_IMPLIB` isn't used.

I'll re-run all tests to confirm the name changes don't break anything.

**Comment by kou** [comment]
@ursabot crossbow submit -g conda -g wheel -g linux

**Comment by kou** [comment]
@ursabot crossbow submit centos-8

**Review by pitrou — APPROVED** [review]
+1. This is a great improvement.

**Pull request closed without merging**